compiler/rustc_middle/src/ty/instance.rs

```diff
@@ -135,14 +135,27 @@ pub enum InstanceDef<'tcx> {
     ///
     /// The `DefId` is for `FnPtr::addr`, the `Ty` is the type `T`.
     FnPtrAddrShim(DefId, Ty<'tcx>),
+
+    /// Typecast shim which replaces the `Self` type with the provided type.
+    /// This is used in vtable calls, where the type of `Self` is abstract as of the time of
+    /// the call.
+    ///
+    /// For `target_instance` which are generatable shims, this is done via inserting a cast at the
+    /// beginning of the shim generated by that instance.
+    /// For `Item`s, we instead build a direct call shim with that cast inserted.
+    CfiShim { target_instance: &'tcx InstanceDef<'tcx>, invoke_ty: Ty<'tcx> },
```

Comment:
Does this mean we can remove the code in the interpreter and codegen that currently does this type adjustments? In the interpreter, that's [here](https://github.com/rust-lang/rust/blob/fc9d1a8133616c4cd6db7c2bcf9f4981e8860be5/compiler/rustc_const_eval/src/interpret/terminator.rs#L857-L872).

If that code is needed then I don't understand what this achieves.

Reply:
That code is still needed at the moment because I am conditionally generating shims - they are only generated when CFI is enabled, since they will result in worse code generation (If it's an item, there's a cast trampoline, and if it's not an item, there's a variant of the shim produced). Even if all calls were inlined, it'd still produce extra bulk as each vtable the type appears in gets its own shim.

I could try testing disabling that code when `.cfi_shims()` is true if you think that would be interesting.

Reply:
Actually, my adjustment to `Virtual` (which is unconditional) might make that work, I'll try it in a bit.